src/lib.rs

```diff
@@ -336,22 +340,84 @@ impl Config {
         assert!(output.ends_with(".a"));
         let lib_name = &output[3..output.len() - 2];
         let dst = self.get_out_dir();
+        let msvc = self.get_target().contains("msvc");
+        let compiler = self.get_compiler();
 
         let mut objects = Vec::new();
         let mut src_dst = Vec::new();
+        let mut dependencies = HashMap::new();
         for file in self.files.iter() {
             let obj = dst.join(file).with_extension("o");
             let obj = if !obj.starts_with(&dst) {
                 dst.join(obj.file_name().unwrap())
             } else {
                 obj
             };
-            fs::create_dir_all(&obj.parent().unwrap()).unwrap();
-            src_dst.push((file.to_path_buf(), obj.clone()));
+
+            let mut input_mtime = get_mtime(&file);
+            let output_mtime = get_mtime(&obj);
+
+            let mut cmd = compiler.to_command();
+            if msvc {
+                cmd.arg("/showIncludes").arg("/E");
+            } else {
+                cmd.arg("-M").arg("-MG");
+            }
+            cmd.arg(&file);
+            println!("running: {:?}", cmd);
+
+            if let Ok(Ok(cpp)) = cmd.output()
+                    .map(|o| String::from_utf8(if msvc {o.stderr} else {o.stdout})) {
+
+                let files = if msvc {
+                    // Output is indented dependency per line, with "Note: ..." prefix 
+                    cpp
+                        .replace("Note: including file:", "")
+                } else {
+                    // Output is make rule with line continuations and multiple files per line
+                    cpp
+                        .replace("\\\n", "") // remove line continuations
+                        .replace("\\ ", "\0") // protect escaped spaces
+                        .replace(" ", "\n") // spaces to new lines
+                        .replace("\0", "\\ ") // bring escaped spaces back
+                };
+
+                input_mtime = files
+                    .lines().skip(1)
+                    .chain(vec![file.to_str().unwrap_or("")])
+                    .filter_map(|s| PathBuf::from(s.trim()).canonicalize().ok())
+                    .map(|f| dependencies
+                        .entry(f.clone())
+                        .or_insert(get_mtime(f))
+                        .clone())
+                    .max()
+                    .unwrap_or(FileTime::zero());
+            }
+
+            // build if a dependency is fresher than the output object
+            if input_mtime >= output_mtime {
```

Comment:
If the `/E` or `-M` execution failed (or our parsing failed), shouldn't `output_mtime` be zero here? That is, we want to always execute this in that case regardless of the state of the input/output.

Similarly below if we failed to learn about dependencies we'll want to avoid printing `cargo:rerun-if-changed` because we don't actually know the dependencies.